nixos/modules/system/boot/stage-1-init.sh

```diff
@@ -370,12 +370,13 @@ mountFS() {
 
     mkdir -p "/mnt-root$mountPoint"
 
-    # For CIFS mounts, retry a few times before giving up.
+    # For ZFS and CIFS mounts, retry a few times before giving up.
     local n=0
     while true; do
         mount "/mnt-root$mountPoint" && break
-        if [ "$fsType" != cifs -o "$n" -ge 10 ]; then fail; break; fi
+        if [ \( "$fsType" != cifs -a "$fsType" != zfs \) -o "$n" -ge 10 ]; then fail; break; fi
```

Comment:
I assume upstream does not have this problem because they use `zpool import` instead of mount: https://github.com/openzfs/zfs/blob/master/contrib/dracut/90zfs/mount-zfs.sh.in#L59

Reply:
No, the use `zpool import` to import/open the pool (like we do) but then they also mount the dataset/filesystem (like we do), see https://github.com/openzfs/zfs/blob/99b281f1ae3833826b4e3cee4126a4ea5e6de987/contrib/dracut/90zfs/mount-zfs.sh.in#L80 and https://github.com/openzfs/zfs/blob/99b281f1ae3833826b4e3cee4126a4ea5e6de987/contrib/dracut/90zfs/zfs-lib.sh.in#L76

Note that upstream doesn't retry mounting like I am doing here, and for sure `mount` returning `Device or resource busy` sporadically while mounting the root filesystem like in #25383 seems like a bug in ZFS (which I'm not sure has been fixed), but I think it doesn't hurt to do what I'm doing: it will fix the boot failure in case the bug resurfaces again, while only making the boot hang 10 seconds later in case there is a real mount problem.

Reply:
Can you also link to the issue in the code, so in future we know why it was added?

Reply:
Done, thanks!